From f5c565eb1174416d2fa35f4309d6929b428e30cc Mon Sep 17 00:00:00 2001
From: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
Date: Sat, 25 Sep 2021 00:12:29 +0300
Subject: [PATCH 029/233] [FIO extras] mx7ulp_com: force board_usb_phy_mode to
 device

Set up a USB OTG port.
Add detecting usb phy mode in order to get a working OTG during early
boot.
Add managing power for a USB OTG port.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
---
 board/ea/mx7ulp_com/mx7ulp_com.c | 54 ++++++++++++++++++++++++++++++++
 1 file changed, 54 insertions(+)

diff --git a/board/ea/mx7ulp_com/mx7ulp_com.c b/board/ea/mx7ulp_com/mx7ulp_com.c
index 672c89075ac..7d866557aab 100644
--- a/board/ea/mx7ulp_com/mx7ulp_com.c
+++ b/board/ea/mx7ulp_com/mx7ulp_com.c
@@ -22,6 +22,8 @@ DECLARE_GLOBAL_DATA_PTR;
 
 #define UART_PAD_CTRL		(PAD_CTL_PUS_UP)
 #define QSPI_PAD_CTRL1		(PAD_CTL_PUS_UP | PAD_CTL_DSE)
+#define OTG_ID_GPIO_PAD_CTRL	(PAD_CTL_IBE_ENABLE)
+#define OTG_PWR_GPIO_PAD_CTRL   (PAD_CTL_OBE_ENABLE)
 
 int dram_init(void)
 {
@@ -79,6 +81,54 @@ int board_qspi_init(void)
 }
 #endif
 
+#ifdef CONFIG_USB_EHCI_MX6
+static iomux_cfg_t const usb_otg1_pads[] = {
+	/* gpio for otgid */
+	MX7ULP_PAD_PTC13__PTC13 | MUX_PAD_CTRL(OTG_ID_GPIO_PAD_CTRL),
+	/* gpio for power en */
+	MX7ULP_PAD_PTE15__PTE15 | MUX_PAD_CTRL(OTG_PWR_GPIO_PAD_CTRL),
+};
+
+#define OTG0_ID_GPIO	IMX_GPIO_NR(3, 13)
+#define OTG0_PWR_EN	IMX_GPIO_NR(5, 15)
+
+static void setup_usb(void)
+{
+	mx7ulp_iomux_setup_multiple_pads(usb_otg1_pads,
+						 ARRAY_SIZE(usb_otg1_pads));
+
+	gpio_request(OTG0_ID_GPIO, "otg_id");
+	gpio_direction_input(OTG0_ID_GPIO);
+}
+
+/* Needs to override the ehci power if controlled by GPIO */
+int board_ehci_power(int port, int on)
+{
+	switch (port) {
+	case 0:
+		if (on)
+			gpio_direction_output(OTG0_PWR_EN, 1);
+		else
+			gpio_direction_output(OTG0_PWR_EN, 0);
+		break;
+	default:
+		printf("MXC USB port %d not yet supported\n", port);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int board_usb_phy_mode(int port)
+{
+	/* The device mode is valid for port 0 only */
+	if (port == 0 && gpio_get_value(OTG0_ID_GPIO))
+		return USB_INIT_DEVICE;
+
+	return USB_INIT_HOST;
+}
+#endif
+
 int board_early_init_f(void)
 {
 	setup_iomux_uart();
@@ -103,6 +153,10 @@ int board_init(void)
 	/* address of boot parameters */
 	gd->bd->bi_boot_params = PHYS_SDRAM + 0x100;
 
+#ifdef CONFIG_USB_EHCI_MX6
+	setup_usb();
+#endif
+
 #ifdef CONFIG_FSL_QSPI
 	board_qspi_init();
 #endif
-- 
2.40.1

